apis/network/v1beta1/openstacknet_types.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OpenStackNetSpec defines the desired state of OpenStackNet
+type OpenStackNetSpec struct {
+	// +kubebuilder:validation:Optional
+	//  List of subnets
+	Subnets []OpenStackSubnet `json:"subnets,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1500
+	// MTU of the network
+	MTU int `json:"mtu"`
+
+	// +kubebuilder:validation:Required
+	// DomainName the name of the domain for this network, usually lower(Name)."OSNetConfig.Spec.DomainName"
+	DomainName string `json:"domainName"`
+}
+
+// OpenStackSubnet type
+type OpenStackSubnet struct {
+
+	// +kubebuilder:validation:Required
+	// Name of the subnet
+	Name string `json:"name"`
+
+	// +kubebuilder:validation:Required
+	// Cidr the cidr to use for this network
+	Cidr string `json:"cidr"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Maximum=4094
+	// Vlan ID
+	Vlan int `json:"vlan,omitempty"`
+
+	// +kubebuilder:validation:Required
+	// AllocationRange a set of IPs for assignment
+	AllocationRange AllocationRange `json:"allocationRange"`
+
+	// +kubebuilder:validation:Optional
+	// Gateway optional gateway for the network
+	Gateway string `json:"gateway,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// Routes, list of networks that should be routed via network gateway.
+	Routes []Route `json:"routes,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// IP Reserverations for Subnet
+	Reservations []IPReservation `json:"reservations,omitempty"`
```

Comment:
could also be a map[string]IPReservation where the key is the hostname of the consumer. would make it easy to validate and also for delete a reservation

Reply:
We can, but as commented earlier I wanted not to make it too complex with details about who reserved an IP.